src/c++/perf_analyzer/test_report_writer.cc

```diff
@@ -0,0 +1,97 @@
+// Copyright 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS"" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <sstream>
+#include "doctest.h"
+#include "report_writer.h"
+
+namespace triton { namespace perfanalyzer {
+
+class TestReportWriter {
+ public:
+  static void WriteGpuMetrics(std::ostream& ofs, pa::PerfStatus& status)
+  {
+    ReportWriter rw{};
+    rw.WriteGpuMetrics(ofs, status);
+  }
+};
+
+TEST_CASE("testing WriteGpuMetrics")
+{
+  // Single Gpu setup
+  PerfStatus s1{};
+  Metrics m1{};
+  m1.gpu_utilization_per_gpu.insert(std::pair<std::string, double>("a", 1.0));
+  m1.gpu_power_usage_per_gpu.insert(std::pair<std::string, double>("a", 2.2));
+  m1.gpu_memory_used_bytes_per_gpu.insert(
+      std::pair<std::string, uint64_t>("a", 3));
+  s1.metrics.push_back(m1);
+
+  SUBCASE("single gpu complete output")
+  {
+    std::ostringstream actual_output{};
```

Comment:
I wanted to ensure the string stream was clear between tests so that there was no pollution of the output. 
If I refactor this out to the higher level, I need to replace this line with a clear statement. 6 of 1, half dozen of the other on this one I think.

Reply:
You won't need to clear the string because each subcase has its own stack.

https://github.com/doctest/doctest/blob/master/doc/markdown/tutorial.md#test-cases-and-subcases

> Subcases can be nested to an arbitrary depth (limited only by your stack size). Each leaf subcase (a subcase that contains no nested subcases) will be executed exactly once on a separate path of execution from any other leaf subcase (so no leaf subcase can interfere with another).

Reply:
Pulled variable out of subcases 👍